herd/libdir/aarch64.cat

```diff
@@ -48,72 +48,142 @@ catdep (* This option says that the cat file computes dependencies *)
 
 include "aarch64hwreqs.cat"
 
-(* Coherence-after *)
+(*** Coherence-after ***)
 let ca = fr | co
 
-(* Observed-by *)
-let obs = ([Exp & M | Imp & T & R]; (rf | ca); [Exp & M | Imp & T & R]) & ext
+(*** TLBI-after, DC-after, IC-after ***)
+include "enumerations.cat"
+with TLBI-after from (all-TLBI-Imp_TTD_R-enums local-hw-reqs)
+with DC-after from (all-DC-Exp_W-enums local-hw-reqs)
+with IC-after from (all-IC-Imp_Instr_R-enum local-hw-reqs)
 
-(* TLBI-after *)
-include "tlbi-after.cat"
-with tlbi-after from (ttd-tlbi-pairs hw-reqs)
+(*** Hazard-ordered-before ***)
+
+(** Explicitly-hazard-ordered-before **)
+let Exp-haz-ob =
+  [Exp & R]; po-loc; [Exp & R]; (ca & ext); [Exp & W]
+
+(** TLBI-ordered-before **)
 
 (* TTD-read-ordered-before *)
 let TTD-read-ordered-before =
-  tlbi-after; [TLBI]; po; [dsb.full]; po; [~(Imp & M)]
-  | (if "ETS2" then tlbi-after; [TLBI]; po; [dsb.full]; po; [Imp & TTD & M] else 0)
+  TLBI-after; [TLBI]; po; [dsb.full]; po; [~(Imp & M)]
+  | TLBI-after; [TLBI]; po; [dsb.full]; po; [CSE]; po; [Imp & M]
+  | (if "ETS2" then TLBI-after; [TLBI]; po; [dsb.full]; po; [Imp & TTD & M] else 0)
 
-(* TLBI-Ordered-Before *)
-let tlbi-ob =
+(* TLBI-ordered-before *)
+let TLBI-ob =
      TTD-read-ordered-before
    | tr-ib^-1; TTD-read-ordered-before & ext
    | po-va-loc; TTD-read-ordered-before & ext
 
-(* TLBI-Coherence-After *)
-let tlbi-ca =
-   [TLBI]; tlbi-after; [Imp & TTD & R]; ca; [W]
+(** IC-ordered-before **)
+
+(* Instr-read-ordered-before *)
+let Instr-read-ordered-before =
+  IC-after; [IC]; po; [dsb.full]; po; [~(Imp & M)]
+  | (if "DIC" then ca else 0)
+
+(* IC-ordered-before *)
+let IC-ob = [Imp & Instr & R]; po; [Imp & Instr & R]; Instr-read-ordered-before
+
+(* Hazard-ordered-before *)
```

Comment:
```suggestion
```

I am not sure, I fully understand the significance of 
```(*** ... ***)```
```(** ... **)```
```(* ... *)```
but Hazard-ordered-before is a comment in line 60 as well

Reply:
another thought is that if we've started needing to define sections, that might mean we need to split into more files. Could we add aarch64vmsa.cat and aarch64ifetch.cat?

Reply:
The intention behind the extra-starred comments is akin to sections. Basically, after rearranging the clauses so that they look more like a flattened tree, it seemed helpful to put those section-like comments so that it were a bit clearer what lower-level definitions are for.

No strong opinions from me on removing the comment at line 90. We could factor out the relevant definitions in separate files; however, my concern would be that the definitions are not **that** modular. In particular, hardware-requirements definitions in `aarch64hwreqs.cat` also feature VMSA/ifetch-related addons, but do not lend themselves to graceful modularization.

(CC'ing @jalglave to highlight the ideas.)